resin/logs.py

```diff
@@ -1,133 +1,190 @@
+    def stop(self, d):
+        reactor.callFromThread(d.addCallback, cbDrop)
+
+
+class Logs(object):
+    """
+    This class implements functions that allow processing logs from device.
 
     """
 
+    subscriptions = defaultdict(list)
+
     def __init__(self):
+        self.base_request = BaseRequest()
         self.config = Config()
         self.device = Device()
+        self.settings = Settings()
+        self.subscription_handler = Subscription()
 
-    def _init_pubnub(func):
-        @wraps(func)
-        def wrapper(self, *args, **kwargs):
-            if not hasattr(self, 'pubnub'):
-                pubnub_key = self.config.get_all()['pubnub']
-                self.pubnub = Pubnub(
-                    publish_key=pubnub_key['publish_key'],
-                    subscribe_key=pubnub_key['subscribe_key']
-                )
-            return func(self, *args, **kwargs)
-        return wrapper
-
-    @_init_pubnub
-    def subscribe(self, uuid, callback, error):
+    def __exit__(self, exc_type, exc_value, traceback):
+        reactor.stop()
+
+    def subscribe(self, uuid, callback, error=None, count=None):
         """
-        This function allows subscribing to device logs.
-        There are fields (`m`, `t`, `s`, `c`) in the output which can be unclear. They stand for:
-            m - The log message itself.
-            t - The log timestamp.
-            s - Is this a system message?
-            c - The id of the service which produced this log (or null if the device does not support multiple containers).
+        Subscribe to device logs.
 
         Args:
             uuid (str): device uuid.
-            callback (function): this callback is called on receiving a message from the channel.
-            error (function): this callback is called on an error event.
-            For more details about callbacks in pubnub subscribe, visit here: https://www.pubnub.com/docs/python/api-reference#subscribe
-
-        Examples:
-            # Define callback and error.
-            >>> def callback(message, channel):
-            ...     print(message)
-            >>> def error(message):
-            ...     print('Error:'+ str(message))
-            >>> Logs.subscribe(uuid=uuid, callback=callback, error=error)
+            callback (function): this callback is called on receiving a message.
+            error (Optional[function]): this callback is called on an error event.
+            count (Optional[int]): number of historical messages to include.
+
+        Returns:
+            dict: a log entry will contain the following keys: `isStdErr, timestamp, message, isSystem, createdAt`.
 
         """
 
-        _print_deprecation_warning()
-        channel = self.get_channel(uuid)
-        self.pubnub.subscribe(channels=channel, callback=callback, error=error)
+        self.device.get(uuid)
+        self.subscriptions[uuid].append(self.subscription_handler.add(uuid, callback, error, count))
+        self.subscription_handler.run()
```

Comment:
This looks like an implementation detail of the Subscription class. Better call it within the `add` method